### PR TITLE
feat(lb) support to expost service loadbalance

### LINKED ISSUE #178: Support to expose service via load balancer
Currently the only option is ingress. Load balancer is useful for deploying on cloud

### PR DESCRIPTION
**Pull request by Ghostbaby**
support to expost service loadbalance

fixed #178

Signed-off-by: zhuhuijun <zhuhuijunzhj@gmail.com>

### CODE REVIEW

File: apis/goharbor.io/v1beta1/harborcluster_webhook.go [hunk 1]
```diff
@@ -208,6 +212,30 @@ func (harborcluster *HarborCluster) validateCache() *field.Error {
...
 
+func (harborcluster *HarborCluster) validateExpose() *field.Error {
+	core := field.NewPath("spec").Child("expose").Child("core")
+
+	if harborcluster.Spec.Expose.Core.LoadBalancer != nil {
+		return required(core.Child("loadbalancer"))
+	}
+
+	if harborcluster.Spec.Expose.Core.Ingress != nil {
+		return required(core.Child("ingress"))
```
Comment (bitsf) [inline apis/goharbor.io/v1beta1/harborcluster_webhook.go:223]: 此处逻辑不太对吧

Comment (Ghostbaby) [reply apis/goharbor.io/v1beta1/harborcluster_webhook.go:223]: 确实有问题，我来修复下。

**Review by cndoit18 — APPROVED** [review]
LGTM

**Review by soulseen — COMMENTED** [review]
/lgtm

File: controllers/goharbor/harbor/loadblance.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+	return PortalLB(lbRes), errors.Wrap(err, "cannot add portal lb service")
+}
+
+func (r *Reconciler) GetPortalLB(ctx context.Context, harbor *goharborv1beta1.Harbor) *corev1.Service {
+	if harbor.Spec.Expose.Portal == nil || harbor.Spec.Expose.Portal.LoadBalancer == nil {
+		return nil
+	}
+
+	if !harbor.Spec.Expose.Notary.LoadBalancer.Enable {
```
Comment (bitsf) [inline controllers/goharbor/harbor/loadblance.go:164]: should be `Spec.Expose.Portal`

File: controllers/goharbor/harbor/loadblance.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type CoreLB graph.Resource
+
+func (r *Reconciler) AddCoreLB(ctx context.Context, harbor *goharborv1beta1.Harbor, core Core, portal Portal) (CoreLB, error) {
+	lbRes, err := r.Controller.AddServiceToManage(ctx, r.GetCoreLB(ctx, harbor), core, portal)
+
+	return CoreLB(lbRes), errors.Wrap(err, "cannot add core lb service")
```
Comment (bitsf) [inline controllers/goharbor/harbor/loadblance.go:22]: should add `if err!= nil`

File: manifests/cluster/deployment.yaml [hunk 4]
```diff
@@ -6839,6 +6839,12 @@ spec:
                         required:
                         - host
                         type: object
+                      loadbalancer:
+                        description: TODO Add supports to ClusterIP, LoadBalancer and NodePort by deploying the nginx component
```
Comment (bitsf) [inline manifests/cluster/deployment.yaml:6843]: remove `TODO`

File: apis/goharbor.io/v1beta1/harbor_types.go [hunk 5]
```diff
@@ -704,6 +707,13 @@ type HarborExposeComponentSpec struct {
 	Ingress *HarborExposeIngressSpec `json:"ingress,omitempty"`
 
 	// TODO Add supports to ClusterIP, LoadBalancer and NodePort by deploying the nginx component
```
Comment (bitsf) [inline apis/goharbor.io/v1beta1/harbor_types.go:709]: remove `TODO`

**Review by steven-zou — APPROVED** [review]
LGTM

File: controllers/goharbor/harbor/loadblance.go (hunk 2 above)
Comment (Ghostbaby) [reply controllers/goharbor/harbor/loadblance.go]: done

File: manifests/cluster/deployment.yaml (hunk 4 above)
Comment (Ghostbaby) [reply manifests/cluster/deployment.yaml:6843]: done

File: apis/goharbor.io/v1beta1/harbor_types.go (hunk 5 above)
Comment (Ghostbaby) [reply apis/goharbor.io/v1beta1/harbor_types.go]: done

**Review by bitsf — APPROVED** [review]

**Pull request merged**